core/src/main/java/org/apache/calcite/rel/core/Filter.java

```diff
@@ -166,13 +166,13 @@ protected boolean digestEquals0(Object obj) {
     }
     Filter o = (Filter) obj;
     return traitSet.equals(o.traitSet)
-        && input.equals(o.input)
+        && input.digestEquals(o.input)
         && condition.equals(o.condition)
         && getRowType().equalsSansFieldNames(o.getRowType());
   }
 
   @API(since = "1.24", status = API.Status.INTERNAL)
   protected int digestHash0() {
-    return Objects.hash(traitSet, input, condition);
+    return Objects.hash(traitSet, input.digestHash(), condition);
```

Comment:
It is weird we have to combine hash code with objects.

Reply:
If we can change RelNode#hashCode(), everything will be normal. I am happy to remove digestHash() if we can remove final from hashCode().

Reply:
I am thinking do we need a utiltiy to combine hash codes and digest hash codes?
So the code can be changed to something like:

`Util.combineHash(traitSet.hashCode(), input.digestHash(), condition.hashCode())`

Reply:
I don't think we need this.